### PR TITLE
[334] move degree type to degree details page

### PR DESCRIPTION
**Pull request by JonF1**
### Context
* https://trello.com/c/50tuG9eo/334-move-degree-type-to-degree-details-page

Changing the user journey so that user selects all options on the /new endpoint and posts params to create where they are saved in the database. Now edit action can be used on the confirmation page and degree record can be updated. 

### Changes proposed in this pull request
New ask and finish endpoints that decide the degree route whether uk or non_uk. 
Updated feature tests to reflect user journey changes
Validations dependent on locale code context

### Guidance to review
Check the review app and create degree. Check in database to see that degree has been saved.
https://dfe-twd-rttd-pr-113.herokuapp.com/

Screenshot 2020-10-26 at 15 55 57

screencapture-localhost-3000-trainees-22-degrees-2020-10-26-15_58_46

screencapture-localhost-3000-trainees-22-degrees-2020-10-26-15_57_33

### CODE REVIEW

**Comment by JonF1** [comment]
[Move degree type to degree details page](https://trello.com/c/50tuG9eo/334-move-degree-type-to-degree-details-page)

**Review by gpeng — CHANGES_REQUESTED** [review]
A few comments about breaking things up and testing. It might be difficult to stick with the happy/sad path feature specs as per our testing guidelines with so many paths through a feature (not so much this one but others are more complex) but we should try I think.

File: app/controllers/trainees/degrees_controller.rb [hunk 1]
```diff
@@ -1,26 +1,38 @@
 module Trainees
   class DegreesController < ApplicationController
-    def new
+    def ask
       @degree = trainee.degrees.build
     end
 
+    def finish
+      @degree = trainee.degrees.build(locale_code_params)
+      if @degree.valid?
```
Comment (aliuk2012) [inline app/controllers/trainees/degrees_controller.rb:9]: Could we update the error messages for this? It current asked the user to select a UK or Non-UK degree option which is not available on the page and may confuse users. It should specify that they select yes or no

File: app/controllers/trainees/degrees_controller.rb [hunk 2]
```diff
@@ -1,26 +1,38 @@
 module Trainees
   class DegreesController < ApplicationController
-    def new
+    def ask
       @degree = trainee.degrees.build
     end
 
+    def finish
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/controllers/trainees/degrees_controller.rb:7]: Plus one on this, let's avoid custom actions if we can.

File: app/controllers/trainees/degrees_controller.rb [hunk 3]
```diff
@@ -1,26 +1,38 @@
...
       else
         render :new
       end
     end
 
     def update
       @degree = trainee.degrees.find(params[:id])
-      degree_details_params = @degree.uk? ? uk_degree_details_params : non_uk_degree_details_params
-      @degree.assign_attributes(degree_details_params)
+      @degree.assign_attributes(degree_params)
```
Comment (aliuk2012) [inline app/controllers/trainees/degrees_controller.rb:35]: ❤️

File: app/views/trainees/degrees/ask.html.erb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+          OpenStruct.new(name: "Yes", value: "uk" ),
+          OpenStruct.new(name: "No", value: "non_uk")
+          ] %>
+
+          <%= f.govuk_collection_radio_buttons :locale_code,
+              question,
+              :value,
+              :name,
+              legend: { text: 'Is this a UK degree?' },
+              hint: { text: '' } %>
```
Comment (aliuk2012) [inline app/views/trainees/degrees/ask.html.erb:21]: remove hint if its not needed

File: app/controllers/trainees/degrees_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (JonF1) [reply app/controllers/trainees/degrees_controller.rb]: Thanks, I've made some changes

File: app/controllers/trainees/degrees_controller.rb (hunk 1 above)
Comment (JonF1) [reply app/controllers/trainees/degrees_controller.rb]: Fixed, cheers

File: app/views/trainees/degrees/ask.html.erb (hunk 4 above)
Comment (JonF1) [reply app/views/trainees/degrees/ask.html.erb]: fixed

File: spec/features/trainees/degrees/asking_degree_spec.rb [hunk 5]
```diff
@@ -1,46 +1,42 @@
 require "rails_helper"
 
-RSpec.feature "Adding a degree" do
-  scenario "without selecting a degree type" do
+RSpec.feature "Selecting the degree route" do
```
[the remark this answers is not in the document]
Comment (JonF1) [reply spec/features/trainees/degrees/asking_degree_spec.rb]: let me know if its ok now 👍

**Review by aliuk2012 — APPROVED** [review]

File: spec/features/trainees/degrees/editing_degree_spec.rb [hunk 6]
```diff
@@ -0,0 +1,101 @@
+require "rails_helper"
+
+RSpec.feature "editing a degree" do
+  context "UK degree" do
+    scenario "filling in the fields correct" do
```
Comment (gpeng) [inline spec/features/trainees/degrees/editing_degree_spec.rb:5]:
```suggestion
    scenario "the user enters valid details" do
```

File: spec/features/trainees/degrees/editing_degree_spec.rb [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      given_a_trainee_with_a_uk_degree
+      when_i_visit_the_edit_degree_details_page
+      and_i_fill_the_uk_degree_details
```
Comment (gpeng) [inline spec/features/trainees/degrees/editing_degree_spec.rb:8]:
```suggestion
      and_i_enter_valid_uk_degree_details
```

File: spec/features/trainees/degrees/editing_degree_spec.rb [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      then_i_click_the_continue_button
+      and_i_am_redirected_to_the_summary_page
+    end
+
+    scenario "seeing validations" do
```
Comment (gpeng) [inline spec/features/trainees/degrees/editing_degree_spec.rb:13]:
```suggestion
    scenario "the user enters invalid details" do
```

Comment (gpeng) [reply spec/features/trainees/degrees/editing_degree_spec.rb:13]: And the same sort of thing in the other two scenarios

**Review by gpeng — CHANGES_REQUESTED** [review]
A couple of text things in the specs. I still don't think we should feature test all paths through the section, especially if we're covering 'the difference' in controller/view specs. It's not much of a problem here as there is only non-uk and uk but there will be more overhead in some of the other sections. Testing either route takes the feature from the user to the database and back and exercises all of the parts so I'd be content with that. Happy to approve this one from that perspective but I'm wary of setting a precedent with this.

**Review by gpeng — APPROVED** [review]

**Pull request merged**